### ISSUE TITLE
Knurling by cutting doesn't work with more complex cutting shapes

### ISSUE THREAD

**Issue by mryndzionek** [CadQuery/cadquery#1442]
I'm trying to do knurling on a cylinder by cutting swept shapes. If I use a shape with a triangular section:
![Screenshot at 2023-11-12 20-07-43](https://github.com/CadQuery/cadquery/assets/786191/1da068a8-0d01-4959-8eeb-d7913c018423)
it works great:
![Screenshot at 2023-11-12 20-08-10](https://github.com/CadQuery/cadquery/assets/786191/72e47000-f452-4267-87ff-15ea0ce4d6e8)

With a rounded section:
![Screenshot at 2023-11-12 20-08-48](https://github.com/CadQuery/cadquery/assets/786191/64c1c392-d432-4870-aa24-95afc694cb59)
it breaks down:
![Screenshot at 2023-11-12 20-09-28](https://github.com/CadQuery/cadquery/assets/786191/ede93ae9-96d0-4358-b3b6-cafd3dcd68ed)
Any hints on how to fix this? I tried cutting in different order, unioning before cutting etc. and it always is wrong.

**Comment by jmwright** [comment]
@mryndzionek Can you check `isValid()` on the rounded shapes before cutting and then after? https://cadquery.readthedocs.io/en/latest/classreference.html?highlight=isvalid#cadquery.Shape.isValid

**Comment by mryndzionek** [comment]
> @mryndzionek Can you check `isValid()` on the rounded shapes before cutting and then after? https://cadquery.readthedocs.io/en/latest/classreference.html?highlight=isvalid#cadquery.Shape.isValid

I've checked `val().isValid()` and it only returns `True`.

**Comment by mryndzionek** [comment]
I've noticed that I even have problem with unioning the two "directions". I can union same "directions":
![Screenshot at 2023-11-13 07-42-20](https://github.com/CadQuery/cadquery/assets/786191/57a750c9-3f7e-4f26-99cd-2317f543a353)
but then when I try unioning those two together it fails. Similarly I can cut those two from a cylinder separately:
![Screenshot at 2023-11-13 07-50-35](https://github.com/CadQuery/cadquery/assets/786191/a0aca012-d429-43bc-afc8-67a605bf63af)
, but when I cut one after another the result is wrong:
![Screenshot at 2023-11-13 07-51-44](https://github.com/CadQuery/cadquery/assets/786191/ac1c859a-7aad-4a0e-9aa1-45112e7abcdd)

**Comment by jmwright** [comment]
You could try sweeping each individual cut (or all the cuts in one direction followed by all the cuts in the other direction) with sweep's `combine` argument set to `cut` to see if that gives better results.

https://github.com/CadQuery/cadquery/blob/7bc8c00dc1a81633daea4810e04762d61a2abd5a/cadquery/cq.py#L3175

**Comment by jmwright** [comment]
@mryndzionek Are you able to post your code?

**Comment by jmwright** [comment]
This issue could also be related. https://github.com/CadQuery/cadquery/issues/1437

**Comment by jmwright** [comment]
I get CAD kernel errors if I make even small changes to your code, so I think you're running in to a kernel issue. I haven't found a magic work-around yet, and CQ-editor crashed my desktop environment on one try. Have you tried unioning all of the cuts together and then cutting the resulting compound at once from the cylinder?

**Comment by mryndzionek** [comment]
> Have you tried unioning all of the cuts together and then cutting the resulting compound at once from the cylinder?

Yes, no major difference, maybe even somewhat worse result. For me CQ-editor deadlocks one some values (like small point radius, etc).

**Comment by jmwright** [comment]
I tried your code in a development environment with OCP/OCCT 7.7.2 and got the same result. Have you tried varying the profile other than triangle vs rounded? For instance, a trapezoid and a trapezoid with rounded corners? I think you are stuck trying to find a combination that that CAD kernel can handle.

**Comment by mryndzionek** [comment]
Trapezoid is also wrong:
![Screenshot at 2023-11-15 21-12-30](https://github.com/CadQuery/cadquery/assets/786191/6a4dc3b3-bb4c-4918-8827-f80842f6ae5a)
The only profile which seems to work is a triangle.

**Comment by jmwright** [comment]
Ok. I can only go up to 0.29 on the knurl cut_radius before the failure happens.
```python
res = knurl(res, height + 1, radius, 50, 0.29, 2, 100, 10)
```
There might be a clever way to use offset twist extrudes with union to achieve the same effect, but I really don't know.

**Comment by lorenzncode** [comment]
Here is an example based on mailing list post:
https://groups.google.com/g/cadquery/c/So2xgLvxmXs/m/WzkdLhN1CAAJ

It doesn't resolve the OCCT issue though it happens to work for these values.
I'm sharing it more as a `Sketch` example.  Warning - the boolean ops are slow here with n=10.

```py
import cadquery as cq
import math

r = 20
h = 50
n = 10  # number of cuts (per direction);
a = 120
cut_angle = 50
cut_depth = 2
length = 2 * cut_depth
point_radius = 0.5

x1 = length * math.cos(math.radians(cut_angle / 2))
y1 = length * math.sin(math.radians(cut_angle / 2))


def doprofile(loc):
    s = (
        cq.Sketch()
        .push([(-cut_depth, 0)])
        .segment((0, 0), (x1, y1))
        .segment((x1, -y1))
        .close()
        .assemble()
        .reset()
        .vertices("<X")
        .fillet(point_radius)
        .reset()
    )
    return s.moved(loc)


s1 = cq.Sketch().parray(r, 0, 360, n).each(doprofile)

w = cq.Workplane()
w1 = w.placeSketch(s1).twistExtrude(h, a)
w2 = w.placeSketch(s1).twistExtrude(h, -a)

tool = w1.union(w2, clean=True, tol=1e-2)
part = w.circle(r).extrude(h).cut(tool, tol=1e-2)
```

![image](https://github.com/CadQuery/cadquery/assets/16394272/b6c63a9f-5a40-4b49-ad41-10eb16475f25)

**Comment by mryndzionek** [comment]
I've tried using `twistExtrude` with the full circular profile and then combine with `intersect`. With `clean=True, tol=1e-2` it seems to be giving good results even for larger `point_radius`es:
![Screenshot at 2023-11-18 22-35-26](https://github.com/CadQuery/cadquery/assets/786191/05d7fc1d-4041-40e1-952a-5c9cecd82110)
